examples/provider_upgrade_test.go

```diff
@@ -0,0 +1,411 @@
+	fixedStateFile := filepath.Join(tempDir, "fixed-state.json")
+	writeFile(t, fixedStateFile, []byte(fixedState))
+	return fixedStateFile
+}
+
+func importState(t *testing.T, pt *integration.ProgramTester, stateFile string) {
+	fixedStateFile := fixupStackName(t, pt, stateFile)
+	pt.RunPulumiCommand("stack", "import", "--file", fixedStateFile)
+}
+
+func parseStackName(t *testing.T, state string) string {
+	type model struct {
+		Deployment struct {
+			Resources []struct {
+				URN  string `json:"urn"`
+				Type string `json:"type"`
+			} `json:"resources"`
+		} `json:"deployment"`
+	}
+	var m model
+	err := json.Unmarshal([]byte(state), &m)
+	require.NoError(t, err)
+	var stackUrn string
+	for _, r := range m.Deployment.Resources {
+		if r.Type == "pulumi:pulumi:Stack" {
+			stackUrn = r.URN
+		}
+	}
+	require.NotEmptyf(t, stackUrn, "failed to find stack URN")
+	parts := strings.Split(strings.TrimPrefix(stackUrn, "urn:pulumi:"), "::")
+	require.NotEmptyf(t, parts, "failed to parse stack URN: %v", stackUrn)
+	require.NotEmptyf(t, parts[0], "found empty stack URN")
+	return parts[0]
+}
+
+func withUpdatedStackName(t *testing.T, newStackName string, state string) string {
+	return strings.ReplaceAll(state, parseStackName(t, state), newStackName)
+}
+
+func deleteFileIfExists(t *testing.T, file string) {
+	err := os.Remove(file)
+	if errors.Is(err, fs.ErrNotExist) {
+		return
+	}
+	require.NoError(t, err)
+}
+
+func writeFile(t *testing.T, file string, data []byte) {
+	ensureFolderExists(t, filepath.Dir(file))
+	err := os.WriteFile(file, data, 0755)
+	require.NoError(t, err)
+}
+
+func readFile(t *testing.T, file string) string {
+	bytes, err := os.ReadFile(file)
+	require.NoError(t, err)
+	return string(bytes)
+}
+
+func ensureFolderExists(t *testing.T, dir string) {
+	if _, err := os.Stat(dir); os.IsNotExist(err) {
+		err = os.MkdirAll(dir, 0755)
+		require.NoError(t, err)
+	}
+}
+
+func parseProviderVersion(t *testing.T, yamlFile string) string {
+	type model struct {
+		Resources struct {
+			Provider struct {
+				Options struct {
+					Version string `yaml:"version"`
+				} `yaml:"options"`
+			} `yaml:"provider"`
+		} `json:"resources"`
+	}
+	bytes, err := os.ReadFile(yamlFile)
+	require.NoError(t, err)
+	var m model
+	yaml.Unmarshal(bytes, &m)
+	require.NoError(t, err)
+	v := m.Resources.Provider.Options.Version
+	require.NotEmptyf(t, v, "Failed to parse Pulumi.yaml: "+
+		"resources.provider.options.version is empty")
+	return v
+}
+
+func ignoreStables(t *testing.T, grpcLogEntry string) string {
+	var v map[string]any
+	err := json.Unmarshal([]byte(grpcLogEntry), &v)
+	require.NoError(t, err)
+	if r, ok := v["response"]; ok {
+		r := r.(map[string]any)
+		if _, ok := r["stables"]; ok {
+			r["stables"] = "*"
+		}
+	}
+	out, err := json.Marshal(v)
+	require.NoError(t, err)
+	return string(out)
+}
+
+func providerServer(t *testing.T) pulumirpc.ResourceProviderServer {
+	ctx := context.Background()
+	version.Version = "0.0.1"
```

Comment:
`version.Version` is a global variable, normally set by the linker. We set it here because we expect that `version.Version` is valid semver (which `""` is not).